rust-alternate/src/connection.rs

```diff
@@ -0,0 +1,256 @@
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use anyhow::{Context as _, Result};
+use futures_util::{
+    stream::{SplitSink, SplitStream},
+    Sink, SinkExt, Stream, StreamExt,
+};
+use tokio::net::TcpStream;
+use tokio_tungstenite::{tungstenite::Message as WebsocketMessage, WebSocketStream};
+
+use crate::{
+    game_result::GameStats,
+    message::{Message, MessageType},
+};
+
+pin_project_lite::pin_project! {
```

Comment:
Hey I went to https://crates.io/crates/pin-project-lite and it assumes I know what the heck is going on here.

What is this?  What does it do?

Reply:
pin-project is useful if you want to create a reference  to something async Rust. Rust requires that any reference used have the Pin trait, so that it can't be moved while the task has yielded or is doing other things. pin-project and pin-project-lite provide convenience macros to Pin things.

https://blog.cloudflare.com/pin-and-unpin-in-rust/
https://docs.rs/pin-project/latest/pin_project/attr.pin_project.html

Reply:
I think this part of code should really be extracted to/replaced with library, i think tokio_codec may be used here (Via its Framed wrappers), because this shouldn't end in user code :D
Pin projections used here to implement Stream/Sink traits for this struct

As general introduction to Pins, and why they are important, you can look at Amos write-up (He also has other great articles)
https://fasterthanli.me/articles/pin-and-suffering

Reply:
blessings.

tons of reading here